Add new version of board beatmap panel

CloneWith opened the pull request:
This version of beatmap panel is mainly designed for the new board.

CloneWith committed:
Add new version of board beatmap panel

NaughtyChas reviewed (commented):
![image](https://github.com/user-attachments/assets/be019d88-fcfc-44d1-8ffa-e8ae58abfb7f)

The animation looks like "no anti-aliasing" on the edges. Fixing this can be considered as impossible, so it's acceptable to leave it as is. 

In accordance with the design, we need to add the visuals for the map ban, and map selected before this can be merged.

cdwcgt commented on osu.Game.Tournament/Components/FumoBeatmapPanel.cs:
```diff
@@ -0,0 +1,337 @@
+                if (!IsLoaded)
+                    return;
+
+                if (value)
+                {
+                    beatmapInfoContainer.BorderColour = Colour4.White;
+                    beatmapInfoContainer.TransformTo(nameof(beatmapInfoContainer.BorderThickness), INNER_BORDER, 500, Easing.OutQuint);
+                    backgroundAddition.FlashColour(Color4.White.Opacity(0.9f), 1000, Easing.OutQuint);
+                }
+                else
+                {
+                    beatmapInfoContainer.TransformTo(nameof(beatmapInfoContainer.BorderThickness), 0f, 500, Easing.OutQuint);
+                }
```
得打包出来
至少在初始化并且在Load或者LoadComplete (我更倾向LoadComplete) 的时候跑一次

不然如果初始化一个Selected = true的FumoBeatmapPanel效果是不会有的

CloneWith committed:
Use a list in TournamentMatch to store chess placements

cdwcgt commented on osu.Game.Tournament/Components/FumoBeatmapPanel.cs:
```diff
@@ -0,0 +1,338 @@
+
+            var newPlacement = currentMatch.Value.ChessPlacements.LastOrDefault(p => p.BeatmapID == Beatmap.Beatmap?.OnlineID);
+
+            string choiceText = newPlacement?.OwnerTeam == TeamColour.Red ? "Red" :
+                newPlacement?.OwnerTeam == TeamColour.Blue ? "Blue" : "Map";
```
读起来不如用switch case(

CloneWith replied:
S1 的遗物）

cdwcgt commented on osu.Game.Tournament/Components/FumoBeatmapPanel.cs:
```diff
@@ -0,0 +1,338 @@
+        private void runAnimation(ColourInfo? colour = null)
+        {
+            // Stop any transform process (if exists) first
+            ClearTransforms();
```
FinishTransforms(true); ?
至少得ClearTransforms(true); 吧
不然其实本身是不动的 动的都是Child

CloneWith replied:
确实）

cdwcgt commented on osu.Game.Tournament/Components/FumoBeatmapPanel.cs:
```diff
@@ -0,0 +1,338 @@
+            // Stop any transform process (if exists) first
+            ClearTransforms();
+
+            ColourInfo useColour = colour ?? Color4.White;
```
欸那你为什么不直接给参数初始化成white?

CloneWith replied:
形参默认值必须是编译时常量😅

cdwcgt replied:
错了哥
![mingfengouo_agadvheaahplavc](https://github.com/user-attachments/assets/f372bd95-8b63-4b32-b055-4f9831e2240c)

cdwcgt reviewed (changes requested):
![mingfengouo_agadzbyaahyxuvc](https://github.com/user-attachments/assets/a3a2c5e0-4609-475f-8101-4bf1601d0877)

CloneWith committed:
Update border status once on load complete

CloneWith committed:
Fade out top mask when new state is unavailable

CloneWith committed:
Use switch case for team text in BoardBeatmapPanel

CloneWith committed:
Change to switch case again in FumoBeatmapPanel

CloneWith committed:
Finish transforms recursively in `runAnimation()`

CloneWith committed:
Merge remote-tracking branch 'origin/feature/board-ui' into feature/map-display

cdwcgt reviewed (approved):
![mingfengouo_agadrq4aaiy-avy (3)](https://github.com/user-attachments/assets/7ebb16a6-d9d2-4795-b3ca-cb904a3cca78)

NaughtyChas committed:
NaughtyChas the Animator

NaughtyChas committed:
Adjust font size for some text

Otherwise your visual will look like padding: 0px;.

NaughtyChas reviewed (approved):
See if the new animation is your type. If there aren't any, I think we can 
 ![git-merge-outline-svgrepo-com](https://github.com/user-attachments/assets/8717055e-b767-4030-850c-21d59b97b98b)

NaughtyChas commented:
Oh, something I did not notice last night:

1. The current version of the design has not implemented the Ban display on the bottom right (that pill), so we should add this in this PR?

Red & blue banned after animation: 
![image](https://github.com/user-attachments/assets/9ac0ffae-cca6-4ce0-a79b-45cb6c2ec085)

---

2. Also there is no way to tell if a map has been picked, so it will no longer be unavailable to the players later in the game. Suggest using the current `Banned` visual scheme (darken without "Ban" pill) for maps that have been played (after Red & blue win), and use the original design for banned maps shown in the design page:

![image](https://github.com/user-attachments/assets/d764cd71-8ab2-400d-aa34-6c6a15d17cf6)

Are these requests aligned with your instincts?

Apologies for any confusion and unclear instructions regarding visual implementations on the design page.

CloneWith commented:
> Are these requests aligned with your instincts?

Seems good. I'll add them later this day.

NaughtyChas committed:
Add ban pill

NaughtyChas commented:
> > Are these requests aligned with your instincts?
> 
> Seems good. I'll add them later this day.

See if I can make your lives easier, so you can focus on more important stuff instead.

#98.

NaughtyChas committed:
Animate it!

NaughtyChas committed:
Change anim starting for topMask

NaughtyChas committed:
Add new animation for winning

NaughtyChas committed:
Update osu.Game.Tournament/Components/FumoBeatmapPanel.cs

Co-authored-by: 复予 <clonewith@qq.com>

NaughtyChas committed:
Wrap trophy into a circle

cdwcgt committed:
remove usage of OfType

CloneWith committed:
Remove unnecessary nullable and cast modifiers

CloneWith committed:
Merge branch 'feature/board-ui' into feature/banpill

cdwcgt committed:
passing ChessPlacement to runAnimation

cdwcgt committed:
remove useless FinishiTransform

cdwcgt committed:
remove transform usage here

cdwcgt committed:
remove usage of currentPlacement

CloneWith committed:
Add function to get team strings in TournamentGame

cdwcgt committed:
use GetTeamColour

CloneWith committed:
Merge remote-tracking branch 'origin/feature/banpill' into feature/banpill

# Conflicts:
#	osu.Game.Tournament/Components/FumoBeatmapPanel.cs

cdwcgt committed:
miss-ed

cdwcgt committed:
switch like

CloneWith committed:
Use `GetTeamString` for team strings in InstructionInfo

CloneWith committed:
Add optional colour fallback for GetTeamColour

CloneWith committed:
Use CircularContainer for bottom right pills

CloneWith committed:
Refactor animation functions in FumoBeatmapPanel

CloneWith committed:
Fix full animation playing condition

cdwcgt committed:
cancel the task before new task (updateState)

CloneWith committed:
Merge pull request #98 from CloneWith/feature/banpill

Refine visuals for picked / banned maps

CloneWith committed:
Fix OnFirstSelected patches

Some of them were overwritten by newer commits.

NaughtyChas commented on osu.Game.Tournament/Components/FumoBeatmapPanel.cs:
```diff
@@ -0,0 +1,457 @@
+            // Always finish transforms first!
+            // Do this at the very beginning of animation.
+            // RunTask must run before FinishTransforms.
+            scheduledFloatingBoxAnimation?.RunTask();
```
This will cause an exception:

`“System.InvalidOperationException”类型的未经处理的异常在 osu.Framework.dll 中发生 : 'Can not run a ScheduledDelegate that has been already Completed.'`

Can be re-created by running through all tests in Components/FumoBeatmapPanel group `Placement`.

Update: it seems that applying any two different states will cause this issue.

CloneWith replied:
Should've been resolved in latest commits.

CloneWith committed:
Re-initialize the floating box schedule before calling

CloneWith committed:
Use non-disposing `Clear` methods in MatchCountdown

We are facing a situation which is way too special. Hope the GC would finally clean them up.

NaughtyChas reviewed (approved):
我看中

CloneWith commented:
> 我看中

中出中
